README.md
# debug-sx

An extension for the tiny and high performant Node.js debugging utility [debug][1].

This documentation uses the original documentation of **debug** in a quoted style, in order to allow distinctness between **debug** and **debug-sx** features.

**debug-sx** can be used instead of debug. It extends the original debug instance with a couple of new features.

## Installation

For Javascript:

```bash
$ npm install debug-sx
```

For Typescript:

```bash
$ npm install debug-sx
$ npm install --save-dev @types/debug
```

## Features

   * ability to print out the **location** where the message is created
   * ability to print out **call stack** of Errors
   * **level based logging**
   * **handlers** for console and files can be used to control which message should be printed out in which style to which target
   *  **improved output format** for better readability
   * absolut **timestamps** which format can be configured
   * high performant implementation prevents lack of speed, especially when no messages are created.
   * usable for Javascript and Typescript
   * special handler for remote debugging

## Chapters in this documentation
* [Features](#features)
* [Usage](#usage)
* [Conventions for namespaces](#conventions-for-namespaces)
* [Wildcards](#wildcards)
* [Environment Variables](#environment-variables)
* [Formatters](#formatters)
* [Browser support](#browser-support)
* [Output streams](#output-streams)
* [Colors](#colors)
* [Handlers](#handlers)
* [Typescript](#typescript)
* [Advanced Example](#advanced-example)
* [Remote platform debugging](#remote-platform-debugging)

## Usage

**Example**:

```js
process.env['DEBUG'] = '*';

const debugsx = require('debug-sx');
const version = '1.0';

const debug = {};
      debug.fine = debugsx('main::FINE');
      debug.info = debugsx('main::INFO');
      debug.warn = debugsx('main::WARN');
      debug.err = debugsx('main::ERR');

const testobject = { name: 'to', value: '1000' };

...
debug.info('Start of application, version %s', version);
debug.info('Test object %o', testobject);
debug.warn('Not implemented yet');
debug.err('There is a problem ...\n%e', new Error());
```
Like the original Node.js module [debug][1], the __DEBUG__ environment variable controls, which namespaces are printed out. The usage of regular expressions allows proper configuration.

Log-Levels are part of the namespace. The last "::" in namespace is used to separate module and level name. 

**Example**: only *INFO*, *WARN* and *ERR* messages should be printed out.  
Set the environment variable before start of application (use command *set* on Windows systems).

`DEBUG="*::INFO, *::WARN, *::ERR"`

... or define this configuration by program before you import **debug-sx**:

```js
process.env['DEBUG'] = '*::INFO, *::WARN, *::ERR';
const debugsx = require('debug-sx');
...
```

## Conventions for namespaces

> If you're using this in one or more of your libraries, you _should_ use the name of your library so that developers may toggle debugging as desired without guessing names. If you have more than one debuggers you _should_ prefix them with your library name and use ":" to separate features. For example "bodyParser" from Connect would then be "connect:bodyParser".

-------------------------------------------------

#### Additional features of debug-sx: ####
Use "::" in namespace to add a **level**. You can name the level as you need.  
Some level names are preconfigured: INFO, WARN, ERR,CONF, FINE, DEB


## Wildcards

> The `*` character may be used as a wildcard. Suppose for example your library has debuggers named "connect:bodyParser", "connect:compress", "connect:session", instead of listing all three with `DEBUG=connect:bodyParser,connect:compress,connect:session`, you may simply do `DEBUG=connect:*`, or to run everything using this module simply use `DEBUG=*`.

> You can also exclude specific debuggers by prefixing them with a "-" character.  For example, `DEBUG=*,-connect:*` would include all debuggers except those starting with "connect:".

## Environment Variables

> When running through Node.js, you can set a few environment variables that will change the behavior of the debug logging:

| Name      | Purpose                                         |
|-----------|-------------------------------------------------|
| `DEBUG`   | Enables/disables specific debugging namespaces. |
| `DEBUG_COLORS`| Whether or not to use colors in the debug output. |
| `DEBUG_DEPTH` | Object inspection depth. |
| `DEBUG_SHOW_HIDDEN` | Shows hidden properties on inspected objects. |

>  __Note:__ The environment variables beginning with `DEBUG_` end up being converted into an Options object that gets used with `%o`/`%O` formatters. 

> See the Node.js documentation for [`util.inspect()`](https://nodejs.org/api/util.html#util_util_inspect_object_options) for the complete list.

---------------------------------------------

#### Additional features of debug-sx: ####

A couple of additional environment variables are possible:

| Name      | Purpose                                         |
|-----------|-------------------------------------------------|
| `DEBUG_WMODULE`   | Fixed minimum width of the module name when printed out. Missing characters will be filled with spaces. |
| `DEBUG_WLEVEL`   | Fixed minimum width of the level name when printed out. Missing characters will be filled with spaces. |
| `DEBUG_WTIMEDIFF`   | Fixed minimum width of the time diff value when printed out. Missing characters will be filled with spaces. |
| `DEBUG_TIME`   | Format of printed out time stamp. The format of the module [dateformat][2] can be used<br>Example: 'ddd, yyyy-mm-dd HH:MM:ss.l'  |
| `DEBUG_PREFIX`   | General prefix used on all messages. |
| `DEBUG_LOCATION`   | Enables/disables specific location output namespaces. |
| `DEBUG_STREAM`   | Default output for console handlers. The values `stdout` oder `stderr` are used. If no value is given, `stderr` is used. |

Object inspect definitions can also be given in a dedicated way to one handler or one debug instance.

For an example see file [example.js][4].

## Formatters
> Debug uses [printf-style](https://wikipedia.org/wiki/Printf_format_string) formatting. Below are the officially supported formatters:

| Formatter | Representation |
|-----------|----------------|
| `%O`      | Pretty-print an Object on multiple lines. |
| `%o`      | Pretty-print an Object all on a single line. |
| `%s`      | String. |
| `%d`      | Number (both integer and float). |
| `%j`      | JSON. Replaced with the string '[Circular]' if the argument contains circular references. |
| `%%`      | Single percent sign ('%'). This does not consume an argument. |

---------------------------------------------

#### Additional features of debug-sx: ####

A couple of additional formatters are available:

| Formatter | Representation |
|-----------|----------------|
| `%e`      | Pretty-print an error object with call stack. If the error object contains the attribute *cause*, the value of *cause* (normally a error object too) is printed as error after a "Caused by:" line. This principle is repeated as long, as the error object has an attribute *cause*. |
| `%l`      | Print the source code location |
| `%f`      | Print the result of a callback function. The callback function must be given as parameter and returns a string. |


### Custom formatters

> You can add custom formatters by extending the `debug.formatters` object. For example, if you wanted to add support for rendering a Buffer as hex with `%h`, you could do something like:

```js
const createDebug = require('debug')
createDebug.formatters.h = (v) => {
    return v.toString('hex')
}

// …elsewhere
const debug = createDebug('foo')
debug('this is hex: %h', new Buffer('hello world'))
//   foo this is hex: 68656c6c6f20776f726c6421 +0ms
```

## Browser support

There is no browser support and therefor Web Inspector Colors are not supported.

## Output streams

> By default `debug` will log to stderr, however this can be configured per-namespace by overriding the `log` method.

---------------------------------------------

#### Additional features of debug-sx: ####

`debug-sx` offers handlers for console and files. See the chapter [Handlers](#handlers).

## Colors

Colors can be configured globally using environment variables, or indivdually on handler. As example it is possible to use colors for console and to avoid colors for log file output.

Colors will be defined by a color record:

```js
{
    namespace: 'main.INFO',
    color: 'blue',
    inverse: true 
}
```
Instead of `namespace`, you can also use `module` or `level` to define which namespace should be covered by this color record. It is also allowed to use regular expressions or wildcards in the string. Any string beginning with `/` and ending with `/` is converted into a regular expression. 

The available colors are:  
*bold*, *italic*, *underline*, *white*, *grey*, *black*,  *blue*, *cyan*, *green*, *magenta*, *red* and *yellow*.

__Hint:__ Hold in mind, that support of color depends on terminal type.

These color records will be combined to an array table, which can be used when a handler is created.

```js
const colorConfig = [
    { level: /DEB*/,   color: 'cyan',   inverse: true },
    { level: /FINE*/, color: 'white',   inverse: true },
    { level: /CONF*/, color: 'magenta', inverse: true },
    { level: /INFO*/, color: 'green',   inverse: true },
    { level: 'WARN',  color: 'yellow',  inverse: true },
    { level: 'ERR',   color: 'red',     inverse: true }
];
```
This example is used as *default color table* when no custom colors are configured.

## Handlers

Handlers are used to direct the messages in different style and visiblity to different targets.

For example, you can print out all messages to a log file without locations and colors and in parallel print out "WARN" and "ERR" level messages to the console with location and in a colored style.

General format rules, for example the timestamp format, cannot be set individually, in order to prevent loss of performance.

If no handler is configured, a default console handler for `stderr` will be used by default.

There are two handler types available:
* console handler (for `stdout` or `stderr`)
* raw console handler for remote debugging (output via `console.log`)
* file handler

Handlers can be added or removed while runtime. But hold in mind, that on each change of handler, all debug instances are checked if they are covered by this change. This may need some amount of time, depending of the quantity of debug instances created up to now.

### Usage

Simple Example for parallel logging to console and file:

```js
const debugsx = require('debug-sx');
const debug = {};
      debug.info = debugsx('main::INFO');

const h1 = debugsx.createConsoleHandler();
const h2 = debugsx.createFileHandler('/tmp/app.log');
debugsx.addHandler(h1, h2);
```
The create functions accept up to four parameters:

| Parameter | Description |
|-----------|----------------|
| First     | filename for `createFileHandler`. |
| First     | output for `createConsoleHandler`, the values `stdout` and `stderr` are allowed. |
| Second     | comma seperated list of enabled namespaces (use prefix '-' to disable namespaces). |
| Third    | comma seperated list of namespaces, where also locations should be printed (use prefix '-' to disable). |
| Fourth    | color table (use undefined if no individually color table is desired).|

## Typescript

Typescript definition file is included in this module.

Example for a typescript source file:


```ts
import * as debugsx from 'debug-sx';

const debug: debugsx.ISimpleLogger = debugsx.createSimpleLogger('main');
const h : debugsx.IHandler = debugsx.createConsoleHandler('stdout', '*');
debugsx.addhandler(h);

debug.info('Start of application');
debug.warn(new Error('This is an exception'));
```
There are three default loggers available:
1)  **ISimpleLogger**  
    Levels: *info*, *warn*
2)  **IDefaultLogger**   
    Levels: *fine*, *config*, *info*, *warn*
3)  **IFullLogger**  
    Levels: *finest*, *finer*, *fine*, *config*, *info*, *warn*, *severe*

You can also define custom levels:
```ts
import * as debugsx from 'debug-sx';

const debug: {
    debx: debug.IDebugger, 
    info: debug.IDebugger, 
    warn: debug.IDebugger 
} = {
    debx: debugsx('main::DEBX'),
    info: debugsx('main::INFO'),
    warn: debugsx('main::WARN')
};

const h : debugsx.IHandler = debugsx.createConsoleHandler('stdout', '*');
debugsx.addhandler(h);

debug.debx('Start of application');
```

Find more information about the interface and the available functions in the file [index.d.ts][3].


## Advanced example ##

Example for parallel logging to console and file:

```js
process.env['DEBUG'] = '*::INFO, *::WARN, *::ERR';
process.env['DEBUG_LOCATION'] = '*::INFO';
process.env['DEBUG_COLORS'] = 'true';
process.env['DEBUG_STREAM'] = 'stdout';
process.env['DEBUG_WMODULE'] = '15';
process.env['DEBUG_WLEVEL'] = '6';
process.env['DEBUG_WTIMEDIFF'] = '6';
process.env['DEBUG_TIME'] = 'ddd, yyyy-mm-dd HH:MM:ss.l';

const debugsx = require('debug-sx');
const debug = {};
      debug.info = debugsx('main::INFO');
      debug.warn = debugsx('main::WARN');

const hc = debugsx.createConsoleHandler('stdout', '*', '-*', []);
const hf = debugsx.createFileHandler(
             '/tmp/app.log',     // file name
             '*::ERR,*::WARN',   // enabled namespaces
             '*::ERR,-*::INFO',  // enabled for printing location
             [
                 { level: 'ERR', color: 'red', inverse: true },
                 { namespace: /.*::WARN/, color: 'blue', inverse: true },
                 { module: 'server*', color: 'magenta' }
             ]);

debugsx.addHandler(hc, hf);

debug.info('Start of application');
debug.warn('Message should be visible in file...');

debugsx.removeHandler(hf);
debug.info('Message only visible on console');

setTimeout(() => { debug.info('end of application') }, 1000);

```

## Remote platform debugging

Debugging on remote platforms may cause problems, if streams to stdout or stderr are not 
redirected to the development platform (see also [https://stackoverflow.com/questions/52940890][5]).

In cases, where `console.log` works, a `RawConsoleHandler` could be used, instead of normal `ConsoleHandler`.  
This handler uses the statement `console.log` instead of writing log messages via streams (for example stdout).

```ts
import * as debugsx from 'debug-sx';

const debug: debugsx.IDefaultLogger = debugsx.createDefaultLogger('main');
const h : debugsx.IHandler = debugsx.createRawConsoleHandler('*');
debugsx.addhandler(h);

debug.info('Start of application');
```

[1]: https://www.npmjs.com/package/debug
[2]: https://www.npmjs.com/package/dateformat
[3]: https://github.com/ManfredSteiner/debug-sx/blob/work/index.d.ts
[4]: https://github.com/ManfredSteiner/debug-sx/blob/work/example.js
[5]: https://stackoverflow.com/questions/52940890



CHANGELOG.md
## 0.2.0 / 2018-10-25

  * New RawConsoleHandler for remote platform debugging  
    see [README.md/remote-platform-debugging](README.md#remote-platform-debugging)
  
## 0.1.0 / 2017-05-22
First useable version with documentation in README.md.
Tests still running again.

## 0.0.10 / 2017-05-18

  * types added (Typescript support)
  
## 0.0.9 / 2017-05-18

  * Initial release



index.js
const debug = require('debug');
const debugOriginal = debug;

const sprintf = require('sprintf-js').sprintf;
const dateFormat = require('dateformat');
const fs = require('fs');
const util = require('util');
const semver = require('semver');
const handler = require('./handler');

module.exports =  debug;
debug.humanize = require('ms');

debug.sprefix = process.env['DEBUG_PREFIX'] || ''; 
debug.modulewidth = +process.env['DEBUG_WMODULE'] || 1;
debug.levelwidth = +process.env['DEBUG_WLEVEL'] || 1;
debug.npattern =  '%-' + debug.modulewidth + 's %-' + debug.levelwidth + 's';
debug.tpattern = process.env['DEBUG_TIME'] || 'HH:MM:ss.l';
debug.dpattern = process.env['DEBUG_WTIMEDIFF'] ? '+%-' + process.env['DEBUG_WTIMEDIFF'] + 's' : '';

if (process.env['DEBUG_COLORS'] === undefined) {
  process.env['DEBUG_COLORS'] = 'true';
}

debug.loggers = [];
debug.handlers = [];
debug.defaultHandler = handler.createConsoleHandler();

debug.formatArgsOriginal = debug.formatArgs;
debug.formatArgs = function formatArgs(args) {
  if (!this.activeHandlers) {
    debug.formatArgsOriginal.call(this, args);
  }
}

debug.log = function (...p) {
  if (!Array.isArray(this.activeHandlers)) {
    // no debug-sx object, use original debug log function
    let logFn = exports.log || console.log.bind(console);
    logFn.apply(debug, p);
    return;
  }

  let s = formatLogData(p);
  let sname = sprintf(debug.npattern, this.module, this.level);
  let stime = debug.tpattern ? dateFormat(this.curr, debug.tpattern) : '';
  let sdiff = debug.dpattern ? sprintf(debug.dpattern, debug.humanize(this.diff)) : '';      
  let prefix = debug.sprefix + stime + ' ' + sdiff + ' ' + sname;
  let eprefix = ' '.repeat(prefix.length);
  let split = s.split('\n');
  let msg;
  if (split.length > 1)
    msg = ' ' + split.join('\n ' + eprefix) + '\n' + eprefix + ' -----------------------------------------------------';
  else
    msg = ' ' + split;
  let msgl;

  for (let ah of this.activeHandlers) {
    if (!ah.enabled) continue;
    if (ah.locationEnabled && !msgl) {
      let location = '\n ' + (new Error).stack.split('\n')[3];;
      msgl = msg + location.split('\n').join('\n' + eprefix);
    }
    ah.write(ah.colorOn);
    ah.write(prefix);
    ah.write(ah.colorOff);
    ah.write(ah.locationEnabled ? msgl : msg);
    ah.write('\n');
  }

}

debug.createDebug = function(namespace, opts) {
  debug.opts = opts;
  let d = debug(namespace);
  delete debug.opts;
  return d;
};

debug.initOriginal = debug.init;
debug.init = function (d) {
  debug.initOriginal(d);
  if (typeof debug.opts === 'object') {
    for (let att in debug.opts) 
      d.inspectOpts[att] = debug.opts[att];
  }
  let i = d.namespace.lastIndexOf('::');
  d.module = i !== -1 ? d.namespace.substr(0, i) : d.namespace;
  d.level = i !== -1 ? d.namespace.substr(i+2) : '';
  d.init = init.bind(d);
  d.init();
  debug.loggers.push(d);
}

init = function () {
  this.activeHandlers = [];
  this.locationEnabled = false;
  this.useColors = true;
  let handlers = debug.handlers.length === 0 ? [ debug.defaultHandler ] : debug.handlers; 
  for (let h of handlers) {
    if (!h.enabled(this.namespace)) continue;
    let locationEnabled = h.locationEnabled(this.namespace);
    let colorCodes;
    if (h.colorTable) 
      colorCodes = h.getColorCodes(this.namespace, this.module, this.level);
    else
      this.useColors = false;

    let ah = new handler.ActiveHandler(h, colorCodes, true, locationEnabled);
    this.locationEnabled = this.locationEnabled || locationEnabled;
    this.activeHandlers.push(ah);
  }
  this.enabled = this.activeHandlers.length > 0;
  
}


debug.createConsoleHandler = handler.createConsoleHandler;
debug.createRawConsoleHandler = handler.createRawConsoleHandler;
debug.createFileHandler = handler.createFileHandler;

debug.addHandler = function (...handler) {
  if (!Array.isArray(handler))
    handler = [ handler ];
  this.handlers = this.handlers.concat(handler);
  for (let d of debug.loggers)
    d.init();
}
 
 
debug.removeHandler = function (handler) {
  let i = this.handlers.findIndex(h => handler === h);
  if (i < 0)
    return false;
  this.handlers.splice(i, 1);
  for (let d of debug.loggers)
    d.init();
  return true;
}

debug.formatters.l = function (v) {
  return '\n    ' + (new Error).stack.split('\n')[6];
};

debug.formatters.e = function (e) {
  if (e instanceof Error) {
    let rv = e.stack || e.message;
    while (e.cause && e.cause instanceof Error) {
      e = e.cause;
      rv += '\nCaused by: ' + e.stack || e.message;
    }
    return rv;
  }
  x = this.inspectOpts;
  return util.inspect(e, x).replace(/\s*\n\s*/g, ' ');
};

debug.formatters.f = function (f) {
  if (typeof f === 'function') {
    return f.call(this);
  }
  return '';
}


// function from log4js / layout.js
function wrapErrorsWithInspect(items) {
  return items.map(function(item) {
    if ((item instanceof Error) && item.stack) {
      return { inspect: function() {
        if (semver.satisfies(process.version, '>=6')) {
          return util.format(item);
        } else {
          return util.format(item) + '\n' + item.stack;
        }
      } };
    } else {
      return item;
    }
  });
}

// function from log4js / layout.js
function formatLogData(logData) {
  var data = logData;
  if (!Array.isArray(data)) {
    var numArgs = arguments.length;
    data = new Array(numArgs);
    for (var i = 0; i < numArgs; i++) {
      data[i] = arguments[i];
    }
  }
  return util.format.apply(util, wrapErrorsWithInspect(data));
}

debug.createSimpleLogger = function (namespace) {
  return {
    info: debug.createDebug(namespace + '::INFO'),
    warn: debug.createDebug(namespace + '::WARN')
  };
}

debug.createDefaultLogger = function (namespace) {
  return  {
    fine:   debug.createDebug(namespace + '::FINE'),
    config: debug.createDebug(namespace + '::CONFIG'),
    info:   debug.createDebug(namespace + '::INFO'),
    warn:   debug.createDebug(namespace + '::WARN'),
  };
}
debug.createFullLogger = function (namespace) {
  return  {
    finest: debug.createDebug(namespace + '::FINEST'),
    finer:  debug.createDebug(namespace + '::FINER'),
    fine:   debug.createDebug(namespace + '::FINE'),
    config: debug.createDebug(namespace + '::CONFIG'),
    info:   debug.createDebug(namespace + '::INFO'),
    warn:   debug.createDebug(namespace + '::WARN'),
    severe: debug.createDebug(namespace + '::SEVERE')
  };
}





handler.js

const fs = require('fs');

// http://en.wikipedia.org/wiki/ANSI_escape_code#graphics
ansiColors = {
  'bold': [1, 22],
  'italic': [3, 23],
  'underline': [4, 24],
  'inverse': [7, 27],
  'white': [37, 39],
  'grey': [90, 39],
  'black': [30, 39],
  'blue': [34, 39],
  'cyan': [36, 39],
  'green': [32, 39],
  'magenta': [35, 39],
  'red': [31, 39],
  'yellow': [33, 39]
};

defaultColorTable = [  
  { level: /DEB.*/, color: 'cyan', inverse: true },
  { level: /FINE.*/, color: 'white', inverse: true },
  { level: 'CONFIG', color: 'magenta', inverse: true },
  { level: 'INFO', color: 'green', inverse: true },
  { level: 'WARN', color: 'yellow', inverse: true },
  { level: 'SEVERE', color: 'red', inverse: true }
];


// export interface IHandler {
//   handler: NodeJS.WritableStream | LogFileStream | ConsoleStream;
//   namespace: string;
//   locationNamespace: string;
//   names: RegExp [],
//   skips: RegExp [],
//   location: { names: RegExp [], skips: RegExp [] },
//   colors: string,
//   colorTable: string
// }
class Handler {
  constructor (hstream, namespaces, locationNamespaces, colors) {
    this.hstream = hstream;
    this.namespaces = namespaces || '*';
    this.locationNamespaces = locationNamespaces || '';
    this.colors = colors || true;
    
    this.names = [];
    this.skips = [];
    this.location = { names: [], skips: [] };

    let split = (namespaces || '').split(/[\s,]+/);
    for (let s of split) {
      if (!s) continue; // ignore empty strings
      let n = s.replace(/\*/g, '.*?');
      if (n[0] === '-') {
        this.skips.push(new RegExp('^' + n.substr(1) + '$'));
      } else {
        this.names.push(new RegExp('^' + n + '$'));
      }
    }

    split = (locationNamespaces || '').split(/[\s,]+/);
    for (let s of split) {
      if (!s) continue; // ignore empty strings
      let n = s.replace(/\*/g, '.*?');
      if (n[0] === '-') {
        this.location.skips.push(new RegExp('^' + n.substr(1) + '$'));
      } else {
        this.location.names.push(new RegExp('^' + n + '$'));
      }
    }
    if (typeof colors === 'boolean' || colors === 'true' || colors ==='false') {
      let c = (colors === true || colors === 'true');
      this.colorTable = c ? createColorTable(defaultColorTable) : undefined;
    } else if (colors)
      this.colorTable = createColorTable(colors);

    this.getColorCodes = getColorCodes.bind(this);  
  }

  end () {
    this.hstream.end();
  }

  enabled (name) {
    for (let re of this.skips) {
      if (re.test(name)) {
        return false;
      }
    }
    for (let re of this.names) {
      if (re.test(name)) {
        return true;
      }
    }
    return false;
  }

  locationEnabled (name) {
    for (let re of this.location.skips) {
      if (re.test(name)) {
        return false;
      }
    }
    for (let re of this.location.names) {
      if (re.test(name)) {
        return true;
      }
    }
    return false;
  }
}


// export interface IActiveHandler {
//   handler: IHandler;
//   colorOn: string;
//   colorOff: string;
//   locationEnabled: boolean;
// }
class ActiveHandler {
  constructor (handler, colorCodes, enabled, locationEnabled) {
    this.handler = handler;
    this.colorOn = colorCodes ? colorCodes.on : '';
    this.colorOff = colorCodes ? colorCodes.off : '';
    this.enabled = enabled;
    this.locationEnabled = locationEnabled;
    this.write = handler.hstream.write.bind(handler.hstream);
  }
}

class HandlerWriteStream {
  constructor (wstream) {
    this.wstream = wstream;
    this.writeable = true;
  }
  
  write (...args) {
    if (this.writeable)
      return this.wstream.write.apply(this.wstream, args);
    else
      return false;
  }

  end () {
    if (this.writeable) {
      this.wstream.end();
      this.writeable = false;
    }
  }
}

class HandlerConsoleStream extends HandlerWriteStream {
  constructor (name) {
    super(process[name]);
    this.name = name;
  }

  end () {
    if (this.writeable) 
      this.writeable = false;
  }
}


class HandlerRawConsole {
  constructor () {
    this.writeable = true;
    this.lineout = '';
  }
  
  write (...args) {
    if (this.writeable) {
      if (Array.isArray(args)) {
        for (const s of args) {
          if (typeof(s) === 'string') {
            if (s !== '') {
              const i = s.indexOf('\n');
              if (i >= 0) {
                this.lineout += s.substr(0, i);
                console.log(this.lineout);
                this.lineout = s.substr(i + 1);
              } else {
                this.lineout += s;
              }
            }
          }
        }
      }
      return true;
    } else {
      return false;
    }
  }

  end () {
    if (this.writeable) {
      this.writeable = false;
    }
  }
}

class HandlerFileStream extends HandlerWriteStream {
  constructor (filename, wstream) {
    super(wstream ? wstream : fs.createWriteStream(filename, {flags: 'a'}));
    if (!wstream) {
      this.wstream.on('finish', function () {
        console.log('log file ' + filename + ' has been written');
      });
      this.wstream.on('error', function (err) {
        console.log('log file ' + filename + ' error');
        console.log(err);
      });
      this.wstream.on('close', function () {
        console.log('log file ' + filename + ' closed');
      });  
    }
    this.filename = filename;
  }
}

function convertPattern (pattern) {
  if (!pattern || (pattern instanceof RegExp))
    return pattern;
  
  if (typeof pattern !== 'string')
    return undefined; // value ignored

  if (pattern.indexOf('*') > 0 || (pattern.startsWith('/') && pattern.endsWith('/')) ) {
    let e = pattern.replace(/\*/g, '.*?');
    if (e.startsWith('/') && e.endsWith('/'))
      e = e.substr(1, e.length-2);
    return new RegExp('^' + e + '$')
  }

  return pattern;
}

function createColorTable (config) {
  let rv = { namespace: {}, level: {}, module: {}, regExpNamespace: [], regExpLevel: [], regExpModule: [] };
  if (!config)
    return rv;
  if (!Array.isArray(config))
    config = [ config ];

  for (let c of config) {
    let r = {};
    r.on = c.inverse ? '\u001b[' + ansiColors['inverse'][0] + 'm' : '';
    r.on += '\u001b[' + ansiColors[c.color][0] + 'm';
    r.off = c.inverse ? '\u001b[' + ansiColors['inverse'][1] + 'm' : '';
    r.off += '\u001b[' + ansiColors[c.color][1] + 'm';
    
    let pL = convertPattern(c.level);
    let pM = convertPattern(c.module);
    let pN = convertPattern(c.namespace);

    if (pN instanceof RegExp)
      r.regExp = pN;
    else if (pL instanceof RegExp)
      r.regExp = pL;
    else if (pM instanceof RegExp)
      r.regExp = pM;

    if (pN instanceof RegExp)
      rv.regExpNamespace.push(r);
    else if (pN) 
      rv.namespace[pN] = r;

    if (pL instanceof RegExp) 
      rv.regExpLevel.push(r);
    else if (pL) 
      rv.level[pL] = r;
    
    if (pM instanceof RegExp) 
      rv.regExpModule.push(r);
    else if (pM) 
      rv.module[pM] = r;
  }
  return rv;
}

function getColorCodes (namespace, module, level) {
    if (!this.colorTable)
      return undefined;
    
    let rv = this.colorTable.namespace[namespace] || this.colorTable.level[level] || this.colorTable.module[module];
    if (rv)
      return rv;
    
    for (let r of this.colorTable.regExpNamespace) {
      if (r.regExp.test(namespace))
        return r;
    }
    for (let r of this.colorTable.regExpLevel) {
      if (r.regExp.test(level))
        return r;
    }
    for (let r of this.colorTable.regExpModule) {
      if (r.regExp.test(module))
        return r;
    }
    
    return undefined;
  }

function createConsoleHandler (fd, namespaces, locationNamespaces, colors) {
  fd = fd || process.env['DEBUG_STREAM'] || 'stderr';
  namespaces = namespaces || process.env['DEBUG'];
  locationNamespaces = locationNamespaces || process.env['DEBUG_LOCATION'];
  colors = colors || process.env['DEBUG_COLORS'];
  let hstream = new HandlerConsoleStream(fd ? fd : 'stderr');
  return new Handler(hstream, namespaces, locationNamespaces, colors);
}

function createRawConsoleHandler (namespaces, locationNamespaces, colors) {
  namespaces = namespaces || process.env['DEBUG'];
  locationNamespaces = locationNamespaces || process.env['DEBUG_LOCATION'];
  colors = colors || process.env['DEBUG_COLORS'];
  let hstream = new HandlerRawConsole();
  return new Handler(hstream, namespaces, locationNamespaces, colors);
}

function createFileHandler (filename, namespaces, locationNamespaces, colors) {
  if (typeof filename === 'object') {
    let config = filename;
    filename = config.filename;
    namespaces = config.namespaces;
    locationNamespaces = config.locationNamespaces;
    colors = config.colors;
  }
  namespaces = namespaces || process.env['DEBUG'];
  locationNamespaces = locationNamespaces || process.env['DEBUG_LOCATION'];
  colors = colors || process.env['DEBUG_COLORS'];
  let hstream = new HandlerFileStream(filename);
  return new Handler(hstream, namespaces, locationNamespaces, colors);
}


module.exports.ActiveHandler = ActiveHandler;
module.exports.createConsoleHandler = createConsoleHandler;
module.exports.createRawConsoleHandler = createRawConsoleHandler;
module.exports.createFileHandler = createFileHandler;

example.js
process.env['DEBUG'] = "*::INFO, *::WARN, *::ERR";
process.env['DEBUG_LOCATION'] = "*::INFO";
process.env['DEBUG_COLORS'] = "true";
process.env['DEBUG_STREAM'] = "stdout";
process.env['DEBUG_WMODULE'] = "15";
process.env['DEBUG_WLEVEL'] = "6";
process.env['DEBUG_WTIMEDIFF'] = "6";
process.env['DEBUG_TIME'] = "ddd, yyyy-mm-dd HH:MM:ss.l";

const debugsx = require('debug-sx');
const version = '1.0';

const debug = {};
      debug.fine = debugsx('main::FINE');
      debug.info = debugsx('main::INFO');
      debug.warn = debugsx('main::WARN');
      debug.err = debugsx('main::ERR');

const testobject = { name: 'to', value: '1000' };

debug.info('Start of application, version %s', version);
debug.fine('Test object %o', testobject);
debug.warn('Not implemented yet');
debug.err('There is a problem ... %e', new Error('Problem 1'));

console.log('Hello world');

setTimeout(() => {
   debug.info('end of application');
}, 1000);

